Implemented unread threads as a relationship on Messagable Trait

DODMax opened the pull request:
Hello,

What do you think of implementing unread messages as a relationship on Messageable Trait?
This way it is possible use it to query the full thread when needed, also to simplify threadsWithNewMessages().

Tests looks ok, let me know if you see any issue I didn't think of.
Cheers

BlueTeck commented:
very nice

cmgmyr commented:
This looks good to me. I've made a handful of changes to `master` a few weeks back (not pushed), so I'll probably have to merge this in locally and make some changes.

cmgmyr commented:
thanks again for the PR!

Gummibeer commented:
since laravel 5.1 `lists()` returns a Collection and not an array. This means that the phpDoc tag is wrong (Depending on laravel version) and the result of the function depends on the framework - not good in my opinion.

Beside this: why does `threadsWithNewMessages()` just returns the IDs? Shouldn't it return the real models and/or if the method should return an array of IDs it should be named `threadIdsWithNewMessages()` or `unreadThreadIds()` - even if I think that plucking the IDs from a EloquentCollection is so simple that it doesn't need an own method!?

DODMax commented:
Indeed I did not think about the recent change in the lists() method.
Maybe we should use something like `->select(Models::table('threads') . '.id')->get()->pluck('id')` (not tested) to avoid this version dependency.

Regarding threadsWithNewMessages() returning IDs I just kept it this way for backward compatibility.

Gummibeer commented:
@DODMax a simple cast to array should solve it - the Collection class has a `toArray()` method which will be called if you cast it to array - if it is already an array nothing will be done - in theory.

DODMax commented:
Seems like I introduced a bug :-/
Messages in threads that were never read (last_read IS NULL) were not showing on some RDBMS.
Fixed in https://github.com/cmgmyr/laravel-messenger/pull/132

cmgmyr commented:
I'm going to revert this commit and revisit these changes shortly
